auto-value-gson/src/test/java/com/ryanharter/auto/value/gson/AutoValueGsonExtensionTest.java

```diff
@@ -809,4 +812,121 @@ public void simple() {
+        + "    private String defaultB = null;\n"
+        + "    private String defaultC = null;\n"
+        + "    private String defaultD = null;\n"
+        + "    public TestTypeAdapter(Gson gson) {\n"
+        + "      this.aAdapter = gson.getAdapter(String.class);\n"
+        + "      this.bAdapter = gson.getAdapter(String.class);\n"
+        + "      this.cAdapter = gson.getAdapter(String.class);\n"
+        + "    }\n"
+        + "    public GsonTypeAdapter setDefaultA(String defaultA) {\n"
+        + "      this.defaultA = defaultA;\n"
+        + "      return this;\n"
+        + "    }\n"
+        + "    public GsonTypeAdapter setDefaultB(String defaultB) {\n"
+        + "      this.defaultB = defaultB;\n"
+        + "      return this;\n"
+        + "    }\n"
+        + "    public GsonTypeAdapter setDefaultC(String defaultC) {\n"
+        + "      this.defaultC = defaultC;\n"
+        + "      return this;\n"
+        + "    }\n"
+        + "    public GsonTypeAdapter setDefaultD(String defaultD) {\n"
+        + "      this.defaultD = defaultD;\n"
+        + "      return this;\n"
+        + "    }\n"
+        + "    @Override\n"
+        + "    public void write(JsonWriter jsonWriter, Test object) throws IOException {\n"
+        + "      if (object == null) {\n"
+        + "        jsonWriter.nullValue();\n"
+        + "        return;\n"
+        + "      }\n"
+        + "      jsonWriter.beginObject();\n"
+        + "      jsonWriter.name(\"a\");\n"
+        + "      aAdapter.write(jsonWriter, object.a());\n"
+        + "      jsonWriter.name(\"b\");\n"
+        + "      bAdapter.write(jsonWriter, object.b());\n"
+        + "      jsonWriter.endObject();\n"
+        + "    }\n"
+        + "    @Override\n"
+        + "    public Test read(JsonReader jsonReader) throws IOException {\n"
+        + "      if (jsonReader.peek() == JsonToken.NULL) {\n"
+        + "        jsonReader.nextNull();\n"
+        + "        return null;\n"
+        + "      }\n"
+        + "      jsonReader.beginObject();\n"
+        + "      String a = this.defaultA;\n"
+        + "      String b = this.defaultB;\n"
+        + "      String c = this.defaultC;\n"
+        + "      String d = this.defaultD;\n"
+        + "      while (jsonReader.hasNext()) {\n"
+        + "        String _name = jsonReader.nextName();\n"
+        + "        switch (_name) {\n"
+        + "          case \"a\": {\n"
+        + "            a = aAdapter.read(jsonReader);\n"
+        + "            break;\n"
+        + "          }\n"
+        + "          case \"c\": {\n"
+        + "            c = cAdapter.read(jsonReader);\n"
+        + "            break;\n"
+        + "          }\n"
+        + "          default: {\n"
+        + "            jsonReader.skipValue();\n"
+        + "          }\n"
+        + "        }\n"
+        + "      }\n"
+        + "      jsonReader.endObject();\n"
+        + "      return new AutoValue_Test(a, b, c, d);\n"
```

Comment:
Hmm, the challenge I'm having with this is that I'd like to avoid a situation where we generate code that is easy to use in a breaking fashion.  In this case, the user consuming this generated TypeAdapter has no way of knowing which default items are required or not, so it's pretty easy to end up in a situation in which your code compiles, runs fine during testing, then a minor change in the JSON response from a server 6 months later will break the code in the wild.  That scares me.

You're right that we can't tell if the default value has been set at every usage during code gen, but we have the other pieces that we can use to make that an unreachable case.  For instance, in the case where an item is ignored for deserialization **and** it's not `@Nullable`, we know this property requires a default value.  Therefore, we could simply add that as a parameter to the constructor of the type adapter, instead of making it a property.  This makes it required and documented (by code), and also means that you can't actually construct a TypeAdapter in a broken state, as opposed to the current code in which `new TestTypeAdapter(gson)` gives you a non-valid (for some responses) object.

Perhaps even logging a warning or info message during generation will help users understand **why** this is happening.

I think this is a good solution, just want to get past that last piece to help users out.